relay-server/src/processing/check_ins/mod.rs

```diff
@@ -0,0 +1,143 @@
+    Processing(#[from] relay_monitors::ProcessCheckInError),
+}
+
+impl OutcomeError for Error {
+    type Error = Self;
+
+    fn consume(self) -> (Option<Outcome>, Self::Error) {
+        let outcome = match &self {
+            Self::RateLimited(limits) => {
+                let reason_code = limits.longest().and_then(|limit| limit.reason_code.clone());
+                Some(Outcome::RateLimited(reason_code))
+            }
+            Self::Processing(relay_monitors::ProcessCheckInError::Json(_)) => {
+                Some(Outcome::Invalid(DiscardReason::InvalidJson))
+            }
+            Self::Processing(_) => Some(Outcome::Invalid(DiscardReason::InvalidCheckIn)),
+        };
+        (outcome, self)
+    }
+}
+
+impl From<RateLimits> for Error {
+    fn from(value: RateLimits) -> Self {
+        Self::RateLimited(value)
+    }
+}
+
+/// A processor for Check-Ins.
+pub struct CheckInsProcessor {
+    limiter: Arc<QuotaRateLimiter>,
+}
+
+impl CheckInsProcessor {
+    /// Creates a new [`Self`].
+    pub fn new(limiter: Arc<QuotaRateLimiter>) -> Self {
+        Self { limiter }
+    }
+}
+
+impl processing::Processor for CheckInsProcessor {
+    type UnitOfWork = SerializedCheckIns;
+    type Output = CheckInsOutput;
+    type Error = Error;
+
+    fn prepare_envelope(
+        &self,
+        envelope: &mut ManagedEnvelope,
+    ) -> Option<Managed<Self::UnitOfWork>> {
+        let headers = envelope.envelope().headers().clone();
+
+        let check_ins = envelope
+            .envelope_mut()
+            .take_items_by(|item| matches!(*item.ty(), ItemType::CheckIn))
+            .into_vec();
+
+        let work = SerializedCheckIns { headers, check_ins };
+        Some(Managed::from_envelope(envelope, work))
+    }
+
+    async fn process(
+        &self,
+        mut check_ins: Managed<Self::UnitOfWork>,
+        ctx: Context<'_>,
+    ) -> Result<Output<Self::Output>, Rejected<Self::Error>> {
+        if ctx.is_processing() {
+            process::normalize(&mut check_ins);
+        }
+
+        self.limiter.enforce_quotas(&mut check_ins, ctx).await?;
+
+        Ok(Output::just(CheckInsOutput(check_ins)))
+    }
+}
+
+/// Output produced by the [`CheckInsProcessor`].
+#[derive(Debug)]
+pub struct CheckInsOutput(Managed<SerializedCheckIns>);
+
+impl Forward for CheckInsOutput {
+    fn serialize_envelope(self) -> Result<Managed<Box<Envelope>>, Rejected<()>> {
+        let envelope = self.0.map(|SerializedCheckIns { headers, check_ins }, _| {
+            Envelope::from_parts(headers, Items::from_vec(check_ins))
+        });
+
+        Ok(envelope)
+    }
+
+    #[cfg(feature = "processing")]
+    fn forward_store(
+        self,
+        s: &relay_system::Addr<crate::services::store::Store>,
+    ) -> Result<(), Rejected<()>> {
+        let envelope = self.serialize_envelope()?;
+        let envelope = ManagedEnvelope::from(envelope).into_processed();
+
+        s.send(crate::services::store::StoreEnvelope { envelope });
```

Comment:
Will we send check ins in a non-envelope store message in the future?

Reply:
Yep! You caught me there, I was a bit lazy and opted to go for the envelope path as everything is already readily available. But longterm I'd like to send a custom message, which also types out the requirement for the `routing_key` (which currently is an item header). Then that is impossible to forget to set and type safe! (Also less random data on `Item` and `EnvelopeHeaders`)